### PR TITLE
Support `List<String>` type for `@Header` annotation

### PR DESCRIPTION
**Pull request by hyangtack**
Motivation:
Several HTTP headers can be sent with the same name. We missed that kind of situation before. Need to enhance annotated HTTP service to support it.

Modifications:
- Support `List<String>` type for `@Header` annotation.

### CODE REVIEW

**Comment by hyangtack** [comment]
This is 1st PR for resolving #1164 . I will send the next PR that `BeanRequestConverterFunction` also supports `List<String>` for `@Header` annotation.

File: core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceTest.java [hunk 1]
```diff
@@ -801,6 +802,11 @@ public void testRequestHeaderInjection() throws Exception {
             request.setHeader("a-name", "minwoox");
             testBody(hc, request, "minwoox is awesome");
 
+            request = post("/11/customHeader");
+            request.addHeader("a-name", "minwoox");
+            request.addHeader("a-name", "giraffe");
+            testBody(hc, request, "minwoox:giraffe is awesome");
```
[the remark this answers is not in the document]
Comment (inch772) [reply core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceTest.java]: image

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java [hunk 2]
```diff
@@ -313,27 +315,65 @@ private HttpResponse convertException(ServiceRequestContext ctx, HttpRequest req
      * Creates a {@link Parameter} instance which describes a parameter of the annotated method. If the
      * parameter type is {@link Optional}, its actual argument's type is used.
      */
-    private static Parameter createOptionalSupportedParam(java.lang.reflect.Parameter parameterInfo,
+    private static Parameter createHttpComponentParameter(java.lang.reflect.Parameter parameterInfo,
                                                           ParameterType paramType, String paramValue) {
-        final Default aDefault = parameterInfo.getAnnotation(Default.class);
-        final boolean isOptionalType = parameterInfo.getType() == Optional.class;
+        assert paramType == ParameterType.PARAM ||
+               paramType == ParameterType.HEADER;
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: How about appending ` : String.valueOf(paramType)`?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java [hunk 3]
```diff
@@ -313,27 +315,65 @@ private HttpResponse convertException(ServiceRequestContext ctx, HttpRequest req
...
+        final boolean isRequired = wrapperType != Optional.class && aDefault == null;
+        return new Parameter(paramType, isRequired,
+                             validateAndNormalizeSupportedType(type), wrapperType,
+                             paramValue, defaultValue, null);
+    }
+
+    private static Class<?> validateWrapperAndElementType(ParameterType paramType,
+                                                          Class<?> clazz, Class<?> elementClazz) {
+        assert paramType == ParameterType.PARAM ||
+               paramType == ParameterType.HEADER;
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: Ditto

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java [hunk 4]
```diff
@@ -313,27 +315,65 @@ private HttpResponse convertException(ServiceRequestContext ctx, HttpRequest req
...
+
+        if (clazz == Optional.class) {
+            return Optional.class;
+        }
+
+        // A list of string is supported only for HTTP headers.
+        if (paramType == ParameterType.HEADER && elementClazz == String.class) {
+            if (clazz == List.class) {
+                return ArrayList.class;
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: Could we support:

- `Iterable` and `List` -> `ArrayList`
- `Set` -> `LinkedHashSet`

..?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java [hunk 5]
```diff
@@ -449,21 +485,42 @@ private static String entryDefaultValue(Parameter entry) {
...
+    private static Object httpHeaderValue(Parameter entry, HttpRequest req) {
         final String name = entry.name();
         assert name != null;
-        final String value = req.headers().get(AsciiString.of(name));
-        if (value != null) {
-            return value;
+
+        final Class<?> wrapperType = entry.wrapperType();
+        if (wrapperType != null && List.class.isAssignableFrom(wrapperType)) {
+            assert entry.type() == String.class;
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: Could we append ` : entry.type().getName()`?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java [hunk 6]
```diff
@@ -449,21 +485,42 @@ private static String entryDefaultValue(Parameter entry) {
...
+            try {
+                @SuppressWarnings("unchecked")
+                final List<String> list = (List<String>) wrapperType.newInstance();
+
+                // We do not call convertParameter() here because the element type is String.
+                final List<String> values = req.headers().getAll(AsciiString.of(name));
+                if (values != null || !values.isEmpty()) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: `&&`?

File: core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceBuilderTest.java [hunk 7]
```diff
@@ -118,6 +121,21 @@ public Object root(@Param("a") byte a) {
                                                  }
                                              },
                                              LoggingService.newDecorator());
+
+        new ServerBuilder().annotatedService(new Object() {
+            @Get("/")
+            public Object root(@Header("a") List<String> a) {
+                return null;
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceBuilderTest.java]: Could use `void` now?

File: core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceBuilderTest.java [hunk 8]
```diff
@@ -118,6 +121,21 @@ public Object root(@Param("a") byte a) {
...
+            }
+        });
+
+        new ServerBuilder().annotatedService(new Object() {
+            @Get("/")
+            public Object root(@Header("a") ArrayList<String> a,
+                               @Header("a") LinkedList<String> b) {
+                return null;
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceBuilderTest.java]: Could use `void` now?

File: core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceBuilderTest.java [hunk 9]
```diff
@@ -193,6 +211,28 @@ public Object root(@Param("name") List<String> name) {
                 return null;
             }
         })).isInstanceOf(IllegalArgumentException.class);
+
+        assertThatThrownBy(() -> new ServerBuilder().annotatedService(new Object() {
+            @Get("/test")
+            public Object root(@Header("name") List<Object> name) {
+                return null;
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceBuilderTest.java]: Could use `void` now?

File: core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceBuilderTest.java [hunk 10]
```diff
@@ -193,6 +211,28 @@ public Object root(@Param("name") List<String> name) {
...
+            }
+        })).isInstanceOf(IllegalArgumentException.class);
+
+        assertThatThrownBy(() -> new ServerBuilder().annotatedService(new Object() {
+            @Get("/test")
+            public Object root(@Header("name") NoDefaultConstructorList<String> name) {
+                return null;
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceBuilderTest.java]: Could use `void` now?

File: core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceTest.java [hunk 11]
```diff
@@ -571,8 +572,8 @@ public String aHeader(@Header String ifMatch) {
         }
 
         @Post("/customHeader")
-        public String customHeader(@Header String aName) {
-            return aName + " is awesome";
+        public String customHeader(@Header List<String> aName) {
+            return Strings.join(aName).with(":") + " is awesome";
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceTest.java]: Could use `String.join()` in JDK

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java (hunk 2 above)
Comment (hyangtack) [reply core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: Sorry, I can't see what you mean. :-( Would you explain why you want to append that?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java (hunk 4 above)
Comment (hyangtack) [reply core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: Ok, let me make a fix to support `Set<String>` as well.

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java (hunk 5 above)
Comment (hyangtack) [reply core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: Do you mean `assert entry.type().getName().equals(String.class.getName())`?

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java (hunk 6 above)
Comment (hyangtack) [reply core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: Oops! Thanks!

File: core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceTest.java (hunk 11 above)
Comment (hyangtack) [reply core/src/test/java/com/linecorp/armeria/server/AnnotatedHttpServiceTest.java]: Oops! Thanks!

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java [hunk 12]
```diff
@@ -318,27 +322,68 @@ private HttpResponse convertException(ServiceRequestContext ctx, HttpRequest req
...
+
+        if (clazz == Optional.class) {
+            return Optional.class;
         }
 
-        return new Parameter(paramType, !isOptionalType && aDefault == null, isOptionalType,
-                             validateAndNormalizeSupportedType(type), paramValue, defaultValue, null);
+        // A list of string is supported only for HTTP headers.
+        if (paramType == ParameterType.HEADER && elementClazz == String.class) {
+            if (clazz == Iterable.class || clazz == List.class) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: I missed the case of `Collection.class`. :sweat_smile:

File: core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java [hunk 13]
```diff
@@ -318,27 +322,68 @@ private HttpResponse convertException(ServiceRequestContext ctx, HttpRequest req
...
+        final Type parameterizedType = parameterInfo.getParameterizedType();
+        if (parameterizedType instanceof ParameterizedType) {
             try {
-                type = (Class<?>) ((ParameterizedType) parameterInfo.getParameterizedType())
-                        .getActualTypeArguments()[0];
-            } catch (Exception e) {
-                throw new IllegalArgumentException("Invalid optional parameter: " + parameterInfo.getName(), e);
+                type = (Class<?>) ((ParameterizedType) parameterizedType).getActualTypeArguments()[0];
+                wrapperType = validateWrapperAndElementType(paramType, parameterInfo.getType(), type);
+            } catch (Throwable cause) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: Do we need to catch here? It seems like `validateWrapperAndElementType` throws same exception with the enough information

Comment (hyangtack) [reply core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java]: You're right. Let me move the line to the out of the try-catch block. Thanks!

**Comment by trustin** [comment]
Thanks, @hyangtack !

**Pull request merged**